### PR TITLE
LibC+Kernel: `termios` yak-shave to get ncurses working through SSH

### LINKED ISSUE #3987: Terminal: cannot scroll or move cursor under less, vim etc when using ssh
I have tested the serenity ssh port now. I was happy to see it let me ssh into one of my VMs and mostly seemed to work but there is at least one showstopper that means ssh isn't going be usable in many cases for interactive terminal work and that is when you open a file under less, vim or what have you you are unable to use the arrow keys to scroll up and down, nor can you move the cursor with the arrow keys under vim.

The arrow keys do work to cycle through command history or edit the command line when using ssh in the serenity terminal but when you open less or vim the arrow keys stop working and the screen just flashes red every time you push them.

### LINKED ISSUE #7235: Ports: nano newline handling broken over SSH
As described previously in #6820, `nano` gets "\r\n" when typing a newline, but it has '\n' bound for a different purpose. Luckily, this is now only present when we ssh into SerenityOS and run nano there.  A quick investigation shows that nano sets different termios settings when running over SSH or running in the Terminal.

Over SSH:
```
[#0 nano(44:44)]: /dev/pts/1 set_termios: ECHO=false, ISIG=true, ICANON=false, ECHOE=true, ECHOK=true, ECHONL=false, ISTRIP=false, ICRNL=false, INLCR=false, IGNCR=false, OPOST=true, ONLCR=true
```
In Terminal:
```
[#0 nano(47:47)]: /dev/pts/0 set_termios: ECHO=false, ISIG=false, ICANON=false, ECHOE=true, ECHOK=true, ECHONL=false, ISTRIP=false, ICRNL=false, INLCR=false, IGNCR=false, OPOST=false, ONLCR=false
```

When running over SSH, the debug console will display warnings about unimplemented `termios` flags, but they are unrelated.

### PR DESCRIPTION
**Pull request by BertalanD**
This PR contains a massive yak-shave with the goal of getting `ncurses` applications working via `ssh` correctly.

Fixes #7235
Fixes #3987

### CODE REVIEW

**Comment by awesomekling** [comment]
The first commit is called "LibC: Remove pointless speed_t to baud rate conversion" but touches a bunch of kernel things.

**Comment by BertalanD** [comment]
Looks like I fat-fingered a git rebase. I'll fix it

File: Userland/Libraries/LibC/termios.cpp [hunk 1]
```diff
@@ -51,79 +51,20 @@ speed_t cfgetospeed(const struct termios* tp)
...
+// POSIX says it's okay if we only fail after calling tcsetattr.
 int cfsetispeed(struct termios* tp, speed_t speed)
 {
-    auto ispeed = baud_rate_from_speed(speed);
-    if (ispeed > 0) {
-        tp->c_ispeed = ispeed;
-    }
-    __RETURN_WITH_ERRNO(ispeed, 0, -1);
+    tp->c_ispeed = speed;
+    errno = 0;
```
Comment (awesomekling) [inline Userland/Libraries/LibC/termios.cpp:59]: POSIX also says that we should not clear `errno` on success, so I think this is wrong.

File: Userland/Libraries/LibC/termios.cpp [hunk 2]
```diff
@@ -51,79 +51,20 @@ speed_t cfgetospeed(const struct termios* tp)
...
 
 int cfsetospeed(struct termios* tp, speed_t speed)
 {
-    auto ospeed = baud_rate_from_speed(speed);
-    if (ospeed > 0) {
-        tp->c_ispeed = ospeed;
-    }
-    __RETURN_WITH_ERRNO(ospeed, 0, -1);
+    tp->c_ospeed = speed;
+    errno = 0;
```
Comment (awesomekling) [inline Userland/Libraries/LibC/termios.cpp:66]: Same here.

File: Userland/Libraries/LibC/sys/ttydefaults.h [hunk 3]
```diff
@@ -46,25 +46,29 @@
...
+        [VTIME] = CTIME,       \
+        [VMIN] = CMIN,         \
+        [VSWTC] = CSWTC,       \
+        [VSTART] = CSTART,     \
+        [VSTOP] = CSTOP,       \
+        [VSUSP] = CSUSP,       \
+        [VEOL] = CEOL,         \
+        [VREPRINT] = CREPRINT, \
+        [VDISCARD]             \
+            = CDISCARD,        \
```
Comment (awesomekling) [inline Userland/Libraries/LibC/sys/ttydefaults.h:64]: Goofy linebreak here.

**Comment by BertalanD** [comment]
I'm sorry that this PR went massively out-of-hand. `termios` is very confusingly documented in the relevant standards, so I can't guarantee that this works correctly. The following tests were performed (both via the Terminal and over SSH) running `cat`:
- Erasing a single character
- Erasing a word
- Erasing an entire line (does not work when the SSH client is running on SerenityOS -- trying to investigate that)
- Erasing multibyte UTF-8 characters

**Comment by danboid** [comment]
Great to see this being worked on!

Should this fix scrolling in vim and less etc?

**Comment by BertalanD** [comment]
Yes, this does fix  `nano`.  The `vim` issues have already been fixed quite some time ago.

The issue was that ncurses tried to set some TTY settings we didn't support, so we returned ENOTIMPL. Because of this, ncurses assumed that we didn't support anything fancy *at all* (basically we were a line printer from the 80s), so didn't even attempt to use the correct escape sequences for cursor movement.

@awesomekling Could you please take a look at this PR?

**Comment by danboid** [comment]
Awesome stuff!

**Comment by BertalanD** [comment]
Commenting to keep this open :shrug:

Will probably need a rebase.

**Pull request closed without merging**